src/main/java/org/jitsi/videobridge/cc/vp8/VP8AdaptiveTrackProjectionContext.java

```diff
@@ -125,330 +96,393 @@ public VP8AdaptiveTrackProjectionContext(
+        @NotNull VideoRtpPacket rtpPacket, int incomingIndex, int targetIndex)
+    {
+        if (!(rtpPacket instanceof Vp8Packet))
         {
-            return null;
+            logger.warn("Packet is not VP8 packet");
+            return false;
         }
 
-        // We know we want to forward this frame, but we need to make sure it's
-        // going to produce a decodable VP8 packet stream.
-        int maxSequenceNumberSeenBeforeFirstPacket
-            = getMaxSequenceNumberOfFrame(
-                rtpPacket.getSsrc(), rtpPacket.getTimestamp());
+        Vp8Packet vp8Packet = (Vp8Packet)rtpPacket;
 
-        VP8FrameProjection nextVP8FrameProjection = lastVP8FrameProjection
-            .makeNext(rtpPacket, maxSequenceNumberSeenBeforeFirstPacket, nowMs);
+        VP8FrameMap.FrameInsertionResult result = insertPacketInMap(vp8Packet);
 
-        if (nextVP8FrameProjection == null)
+        if (result == null)
         {
-            return null;
+            /* Very old frame, more than Vp8FrameMap.FRAME_MAP_SIZE old. */
+            return false;
         }
 
-        if (lastVP8FrameProjection.getSSRC() != nextVP8FrameProjection.getSSRC())
+        VP8Frame frame = result.getFrame();
+
+        if (result.isNewFrame())
         {
-            /* We've changed the encoding we're forwarding, so previous frames
-             * in the map are no longer relevant.
+            if (vp8Packet.isKeyframe() && frameIsNewSsrc(frame))
+            {
+            /* If we're not currently projecting this SSRC, make sure we haven't
+               already decided not to accept a subsequent TL0 frame of this SSRC.
+               If we have, we can't turn on the encoding starting from this
+               packet, so treat this frame as though it weren't a keyframe.
              */
-            vp8FrameProjectionMap.clear();
-        }
-        else
-        {
-            /* Remove projections of old frames from the map. */
-            cleanupFrameProjectionMap(rtpPacket.getTimestamp());
-        }
+                VP8Frame f = findNextTl0(frame);
+                if (f != null && !f.isAccepted())
+                {
+                    frame.setKeyframe(false);
+                }
+            }
+
+            long nowMs = System.currentTimeMillis();
+            boolean accepted = vp8QualityFilter
+                .acceptFrame(frame, incomingIndex, targetIndex, nowMs);
+            frame.setAccepted(accepted);
 
-        // We have successfully projected the incoming frame and we've allocated
-        // a starting sequence number for it. Any previous frames can no longer
-        // grow.
-        vp8FrameProjectionMap.put(rtpPacket.getTimestamp(), nextVP8FrameProjection);
-        // The frame attached to the "last" projection is no longer the "last".
-        lastVP8FrameProjection = nextVP8FrameProjection;
+            if (accepted)
+            {
+                VP8FrameProjection projection = createProjection(frame, vp8Packet);
```

Comment:
Is `createProjection` here doing more of a `getOrCreateProjection`?

Reply:
No, this is in the `isNewFrame` clause.